android-targetSdkVersion preference not honoured

# Bug Report

## Problem

Using the `android-targetSdkVersion` [preference](https://cordova.apache.org/docs/en/latest/config_ref/index.html#preference) does not work.

### What is expected to happen?
The project should use the desired android sdk version as defined in the preference, or fallback to the platform default if not set.


### What does actually happen?
The preference is not considered, and the target sdk always gets set to `28`.


## Information
<!-- Include all relevant information that might help understand and reproduce the problem -->
The code that appears to gets the desired target can be found here: 
https://github.com/apache/cordova-android/blob/c35a990c09f1ab278873754eb7f81e52250893a3/bin/templates/cordova/lib/check_reqs.js#L50-L70

Originally discovered after I suggested to use the `android-targetSdkVersion` preference on https://github.com/apache/cordova-android/issues/830

### Command or Code
<!-- What command or code is needed to reproduce the problem? -->

This is the full `config.xml`:
```xml
<?xml version='1.0' encoding='utf-8'?>
<widget id="com.breautek.cordova.api29" version="1.0.0" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
    <name>HelloCordova</name>
    <description>
        A sample Apache Cordova application that responds to the deviceready event.
    </description>
    <author email="dev@cordova.apache.org" href="http://cordova.io">
        Apache Cordova Team
    </author>
    <content src="index.html" />
    <plugin name="cordova-plugin-whitelist" spec="1" />
    <access origin="*" />
    <allow-intent href="http://*/*" />
    <allow-intent href="https://*/*" />
    <allow-intent href="tel:*" />
    <allow-intent href="sms:*" />
    <allow-intent href="mailto:*" />
    <allow-intent href="geo:*" />
    <platform name="android">
        <allow-intent href="market:*" />
    </platform>
    <platform name="ios">
        <allow-intent href="itms:*" />
        <allow-intent href="itms-apps:*" />
    </platform>
    <preference name="android-targetSdkVersion" value="29" />
</widget>
```

### Environment, Platform, Device
<!-- In what environment, on what platform or on which device are you experiencing the issue? -->
Ubuntu
Android


### Version information
<!-- 
What are relevant versions you are using?
For example:
Cordova: Cordova CLI, Cordova Platforms, Cordova Plugins 
Other Frameworks: Ionic Framework and CLI version
Operating System, Android Studio, Xcode etc.
-->
Cordova `9.0.0 (cordova-lib@9.0.1)`
cordova-android `8.1.0`
Ubuntu `18.04`


## Checklist
<!-- Please check the boxes by putting an x in the [ ] like so: [x] -->

- [x] I searched for existing GitHub issues
- [x] I updated all Cordova tooling to most recent version
- [x] I included all the necessary information above

Sadly, still not working for me even with that preference inside platform:

```console
Checking the license for package Android SDK Platform 28 in /opt/android-sdk-linux/licenses
License for package Android SDK Platform 28 accepted.
Preparing "Install Android SDK Platform 28 (revision: 6)".
Warning: Failed to read or create install properties file.

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> Failed to install the following SDK components:
      platforms;android-28 Android SDK Platform 28
  Install the missing components using the SDK manager in Android Studio.


* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3m 22s
Command finished with error code 1: /app/platforms/android/gradlew :app:bundleRelease,-b,/app/platforms/android/build.gradle
/app/platforms/android/gradlew: Command failed with exit code 1 Error output:
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> Failed to install the following SDK components:
      platforms;android-28 Android SDK Platform 28
  Install the missing components using the SDK manager in Android Studio.


* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

```

My config.xml:

```xml
.....
<platform name="android">
        <preference name="android-minSdkVersion" value="19" />
        <preference name="android-targetSdkVersion" value="29" />
        <config-file parent="/manifest/application/" target="app/src/main/AndroidManifest.xml" xmlns:android="http://schemas.android.com/apk/res/android">
            <meta-data android:name="com.google.firebase.messaging.default_notification_icon" android:resource="@drawable/ic_notification" />
        </config-file>
        <allow-intent href="market:*" />
        <resource-file src="google-services.json" target="google-services.json" />
        <resource-file src="resources/android/icon/drawable-mdpi-icon.png" 
        ....
    </platform>
....
```

--- a/bin/templates/cordova/lib/check_reqs.js
+++ b/bin/templates/cordova/lib/check_reqs.js
@@ -26,10 +26,9 @@ var fs = require('fs');
 var os = require('os');
 var REPO_ROOT = path.join(__dirname, '..', '..', '..', '..');
 var PROJECT_ROOT = path.join(__dirname, '..', '..');
-var events = require('cordova-common').events;
-var CordovaError = require('cordova-common').CordovaError;
-var ConfigParser = require('cordova-common').ConfigParser;
+const { CordovaError, ConfigParser, events } = require('cordova-common');
 var android_sdk = require('./android_sdk');
+const { createEditor } = require('properties-parser');
 
 function forgivingWhichSync (cmd) {
     try {
@@ -53,46 +52,34 @@ module.exports.isDarwin = function () {
  * @returns {string} The android target in format "android-${target}"
  */
 module.exports.get_target = function () {
-    function extractFromFile (filePath) {
-        var target = shelljs.grep(/\btarget=/, filePath);
-        if (!target) {
-            return null;
-        }
-        return target.split('=')[1].trim();
+    const projectPropertiesPaths = [
+        path.join(REPO_ROOT, 'framework', 'project.properties'),
+        path.join(PROJECT_ROOT, 'project.properties')
+    ];
+
+    // Get the minimum required target API from the framework.
+    let target = projectPropertiesPaths.filter(filePath => fs.existsSync(filePath))
+        .map(filePath => createEditor(filePath).get('target'))
+        .pop();
+
+    if (!target) {
+        throw new Error(`We could not locate the target from the "project.properties" at either "${projectPropertiesPaths.join('", "')}".`);
     }
 
-    var target = null;
-    var repo_file = path.join(REPO_ROOT, 'framework', 'project.properties');
-    var project_file = path.join(PROJECT_ROOT, 'project.properties');
-    var config_file = path.join(REPO_ROOT, 'config.xml');
+    // If the repo config.xml file exists, find the desired targetSdkVersion.
+    const configFile = path.join(REPO_ROOT, 'config.xml');
+    if (!fs.existsSync(configFile)) return target;
 
-    // First get the desired target API from the framework. This will be treated as our
-    // minimum required target API.
-    if (fs.existsSync(repo_file)) {
-        target = extractFromFile(repo_file);
-    } else if (fs.existsSync(project_file)) {
-        // if no target found, we're probably in a project and project.properties is in PROJECT_ROOT.
-        target = extractFromFile(project_file);
-    }
+    const configParser = new ConfigParser(configFile);
+    const desiredAPI = parseInt(configParser.getPreference('android-targetSdkVersion', 'android'), 10);
 
-    if (target === null) {
-        throw new Error('We could not locate the "project.properties" at either ' + repo_file + ' or ' + project_file + '.');
-    }
+    if (!isNaN(desiredAPI)) {
+        const minimumAPI = parseInt(target.split('-').pop(), 10);
 
-    var minimumAPI = parseInt(target.split('-')[1]);
-
-    // Next, check the app config.xml file for the desired targetSdkVersion and use it if
-    // found and valid.
-    if (fs.existsSync(config_file)) {
-        // Find the desired android target from the project config
-        var configParser = new ConfigParser(config_file);
-        var desiredAPI = parseInt(configParser.getPreference('android-targetSdkVersion', 'android'));
-        if (!isNaN(desiredAPI)) {
-            if (desiredAPI >= minimumAPI) {
-                target = 'android-' + desiredAPI;
-            } else {
-                events.emit('warn', 'android-targetSdkVersion must be greater than or equal to ' + minimumAPI + '.');
-            }
+        if (desiredAPI >= minimumAPI) {
+            target = `android-${desiredAPI}`;
+        } else {
+            events.emit('warn', `android-targetSdkVersion should be greater than or equal to ${minimumAPI}.`);
         }
     }
 

--- a/bin/templates/cordova/lib/check_reqs.js
+++ b/bin/templates/cordova/lib/check_reqs.js
@@ -26,7 +26,9 @@ var fs = require('fs');
 var os = require('os');
 var REPO_ROOT = path.join(__dirname, '..', '..', '..', '..');
 var PROJECT_ROOT = path.join(__dirname, '..', '..');
+var events = require('cordova-common').events;
 var CordovaError = require('cordova-common').CordovaError;
+var ConfigParser = require('cordova-common').ConfigParser;
 var android_sdk = require('./android_sdk');
 
 function forgivingWhichSync (cmd) {
@@ -55,16 +57,40 @@ module.exports.get_target = function () {
         }
         return target.split('=')[1].trim();
     }
+
+    var target = null;
+
     var repo_file = path.join(REPO_ROOT, 'framework', 'project.properties');
-    if (fs.existsSync(repo_file)) {
-        return extractFromFile(repo_file);
-    }
     var project_file = path.join(PROJECT_ROOT, 'project.properties');
-    if (fs.existsSync(project_file)) {
+    var config_file = path.join(REPO_ROOT, 'config.xml');
+
+    if (fs.existsSync(repo_file)) {
+        target = extractFromFile(repo_file);
+    } else if (fs.existsSync(project_file)) {
         // if no target found, we're probably in a project and project.properties is in PROJECT_ROOT.
-        return extractFromFile(project_file);
+        target = extractFromFile(project_file);
     }
-    throw new Error('Could not find android target in either ' + repo_file + ' nor ' + project_file);
+
+    if (target === null) {
+        throw new Error('Could not find android target in either ' + repo_file + ' nor ' + project_file);
+    }
+
+    var minimumAPI = parseInt(target.split('-')[1]);
+
+    if (fs.existsSync(config_file)) {
+        // Find the desired android target from the project config
+        var configParser = new ConfigParser(config_file);
+        var desiredAPI = parseInt(configParser.getPreference('android-targetSdkVersion', 'android'));
+        if (!isNaN(desiredAPI)) {
+            if (desiredAPI >= minimumAPI) {
+                target = 'android-' + desiredAPI;
+            } else {
+                events.emit('warn', 'android-targetSdkVersion must be greater than or equal to ' + minimumAPI + '.');
+            }
+        }
+    }
+
+    return target;
 };
 
 // Returns a promise. Called only by build and clean commands.

--- a/bin/templates/cordova/lib/check_reqs.js
+++ b/bin/templates/cordova/lib/check_reqs.js
@@ -47,23 +47,27 @@ module.exports.isDarwin = function () {
     return (os.platform() === 'darwin');
 };
 
-// Get valid target from framework/project.properties if run from this repo
-// Otherwise get target from project.properties file within a generated cordova-android project
+/**
+ * @description Get valid target from framework/project.properties if run from this repo
+ *              Otherwise get target from project.properties file within a generated cordova-android project
+ * @returns {string} The android target in format "android-${target}"
+ */
 module.exports.get_target = function () {
     function extractFromFile (filePath) {
         var target = shelljs.grep(/\btarget=/, filePath);
         if (!target) {
-            throw new Error('Could not find android target within: ' + filePath);
+            return null;
         }
         return target.split('=')[1].trim();
     }
 
     var target = null;
-
     var repo_file = path.join(REPO_ROOT, 'framework', 'project.properties');
     var project_file = path.join(PROJECT_ROOT, 'project.properties');
     var config_file = path.join(REPO_ROOT, 'config.xml');
 
+    // First get the desired target API from the framework. This will be treated as our
+    // minimum required target API.
     if (fs.existsSync(repo_file)) {
         target = extractFromFile(repo_file);
     } else if (fs.existsSync(project_file)) {
@@ -72,11 +76,13 @@ module.exports.get_target = function () {
     }
 
     if (target === null) {
-        throw new Error('Could not find android target in either ' + repo_file + ' nor ' + project_file);
+        throw new Error('We could not locate the "project.properties" at either ' + repo_file + ' or ' + project_file + '.');
     }
 
     var minimumAPI = parseInt(target.split('-')[1]);
 
+    // Next, check the app config.xml file for the desired targetSdkVersion and use it if
+    // found and valid.
     if (fs.existsSync(config_file)) {
         // Find the desired android target from the project config
         var configParser = new ConfigParser(config_file);

--- a/spec/unit/check_reqs.spec.js
+++ b/spec/unit/check_reqs.spec.js
@@ -17,11 +17,16 @@
     under the License.
 */
 
-var check_reqs = require('../../bin/templates/cordova/lib/check_reqs');
+var rewire = require('rewire');
+var check_reqs = rewire('../../bin/templates/cordova/lib/check_reqs');
 var android_sdk = require('../../bin/templates/cordova/lib/android_sdk');
 var shelljs = require('shelljs');
 var fs = require('fs');
 var path = require('path');
+var events = require('cordova-common').events;
+
+// This should match /bin/templates/project/build.gradle
+const DEFAULT_TARGET_API = 28;
 
 describe('check_reqs', function () {
     var original_env;
@@ -176,13 +181,87 @@ describe('check_reqs', function () {
             });
         });
     });
+
     describe('get_target', function () {
+        var ConfigParser;
+        var getPreferenceSpy;
+        beforeEach(function () {
+            getPreferenceSpy = jasmine.createSpy();
+            ConfigParser = jasmine.createSpy().and.returnValue({
+                getPreference: getPreferenceSpy
+            });
+            check_reqs.__set__('ConfigParser', ConfigParser);
+        });
+
         it('should retrieve target from framework project.properties file', function () {
             var target = check_reqs.get_target();
             expect(target).toBeDefined();
-            expect(target).toContain('android-');
+            expect(target).toContain('android-' + DEFAULT_TARGET_API);
+        });
+
+        it('should throw error if target cannot be found', function () {
+            spyOn(fs, 'existsSync').and.returnValue(false);
+            expect(function () {
+                check_reqs.get_target();
+            }).toThrow();
+        });
+
+        it('should override target from config.xml preference', () => {
+            var realExistsSync = fs.existsSync;
+            spyOn(fs, 'existsSync').and.callFake(function (path) {
+                if (path.indexOf('config.xml') > -1) {
+                    return true;
+                } else {
+                    return realExistsSync.call(fs, path);
+                }
+            });
+
+            getPreferenceSpy.and.returnValue(DEFAULT_TARGET_API + 1);
+
+            var target = check_reqs.get_target();
+
+            expect(getPreferenceSpy).toHaveBeenCalledWith('android-targetSdkVersion', 'android');
+            expect(target).toBe('android-' + (DEFAULT_TARGET_API + 1));
+        });
+
+        it('should fallback to default target if config.xml has invalid preference', () => {
+            var realExistsSync = fs.existsSync;
+            spyOn(fs, 'existsSync').and.callFake(function (path) {
+                if (path.indexOf('config.xml') > -1) {
+                    return true;
+                } else {
+                    return realExistsSync.call(fs, path);
+                }
+            });
+
+            getPreferenceSpy.and.returnValue(NaN);
+
+            var target = check_reqs.get_target();
+
+            expect(getPreferenceSpy).toHaveBeenCalledWith('android-targetSdkVersion', 'android');
+            expect(target).toBe('android-' + DEFAULT_TARGET_API);
+        });
+
+        it('should warn if target sdk preference is lower than the minimum required target SDK', () => {
+            var realExistsSync = fs.existsSync;
+            spyOn(fs, 'existsSync').and.callFake(function (path) {
+                if (path.indexOf('config.xml') > -1) {
+                    return true;
+                } else {
+                    return realExistsSync.call(fs, path);
+                }
+            });
+
+            getPreferenceSpy.and.returnValue(DEFAULT_TARGET_API - 1);
+
+            var target = check_reqs.get_target();
+
+            expect(getPreferenceSpy).toHaveBeenCalledWith('android-targetSdkVersion', 'android');
+            expect(target).toBe('android-' + DEFAULT_TARGET_API);
+            expect(events.emit).toHaveBeenCalledWith('warn', 'android-targetSdkVersion must be greater than or equal to ' + DEFAULT_TARGET_API + '.');
         });
     });
+
     describe('check_android_target', function () {
         it('should should return full list of supported targets if there is a match to ideal api level', () => {
             var fake_targets = ['you are my fire', 'my one desire'];

--- a/spec/unit/check_reqs.spec.js
+++ b/spec/unit/check_reqs.spec.js
@@ -26,7 +26,7 @@ var path = require('path');
 var events = require('cordova-common').events;
 
 // This should match /bin/templates/project/build.gradle
-const DEFAULT_TARGET_API = 28;
+const DEFAULT_TARGET_API = 29;
 
 describe('check_reqs', function () {
     var original_env;
@@ -258,7 +258,7 @@ describe('check_reqs', function () {
 
             expect(getPreferenceSpy).toHaveBeenCalledWith('android-targetSdkVersion', 'android');
             expect(target).toBe('android-' + DEFAULT_TARGET_API);
-            expect(events.emit).toHaveBeenCalledWith('warn', 'android-targetSdkVersion must be greater than or equal to ' + DEFAULT_TARGET_API + '.');
+            expect(events.emit).toHaveBeenCalledWith('warn', 'android-targetSdkVersion should be greater than or equal to ' + DEFAULT_TARGET_API + '.');
         });
     });
 

